app/services/okta.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+class Okta
+  include HTTParty
+  LEEWAY = 0
+  base_uri ENV.fetch("OKTA_SERVER_URL")
+
+  class << self
+    def find_user_by_id_token(id_token)
+      info = validate_okta_id_token(id_token)
+      find_user(info) || create_user(info)
+    rescue JWT::DecodeError => e
+      raise Okta::FailedAuthentication, e.message
+    end
+
+    private
+
+    def find_user(user_info)
+      User.find_by(okta_id: user_info[:okta_id])
+    end
+
+    def create_user(user_info)
+      User.create!(user_info)
+    end
+
+    def validate_okta_id_token(id_token)
+      transform_jwt_payload JWT.decode(id_token, nil, true, jwt_options).first
+    end
+
+    def transform_jwt_payload(payload)
+      {
+        okta_id: payload["sub"],
+        email: payload["email"],
+        first_name: payload["given_name"],
+        last_name: payload["family_name"]
+      }.with_indifferent_access
+    end
+
+    # based on https://github.com/CruGlobal/omniauth-oktaoauth/blob/e78ffb420cb9ab89fd8951d9fe8a20f175c41618/lib/omniauth/strategies/oktaoauth.rb#L97-L108
+    def jwt_options
+      {
+        algorithm: "RS256",
+        jwks: fetch_jwks,
+        verify_iss: true,
+        verify_aud: true,
+        iss: ENV["OKTA_SERVER_PATH"],
+        aud: ENV["OKTA_SERVER_AUDIENCE"].split(","),
+        verify_expiration: true,
+        verify_not_before: true,
+        verify_iat: true,
+        verify_jti: false,
+        leeway: LEEWAY
+      }
+    end
+
+    def fetch_jwks
+      path = if ENV["OKTA_SERVER_ID"].present?
+        "/oauth2/#{ENV["OKTA_SERVER_ID"]}/v1/keys"
+      else
+        "/oauth2/v1/keys"
+      end
+      response = get(path)
+      JSON.parse(response.body, symbolize_names: true)
+    end
+  end
```

Comment:
I see this method constructing a path, then getting the path, and returning the response parsed by JSON to give us the JWKs. What are the JWKs and what are they for?

Reply:
https://openid.net/specs/draft-jones-json-web-key-03.html
JWK (JSON web key) is a standard on how to structure public signing keys so clients know how to read them.
We use this public signing key (Okta's key) to confirm that the id_token was definitely made by our okta server.

I don't fully understand the internals public-private signing, but it's somewhat magical. Using a private key only the system with the private key can sign (or encrypt) the data, but anyone with the public key can confirm the paired private signed it.

Reply:
I dont fully comprehend this either but I think I get the gist. The public-private keys are used for signing to make sure the data is encrypted and unencrypted between our server and the intended recipients?

Reply:
yeah, basically.